drv/lpc55-swd/src/main.rs

```diff
@@ -0,0 +1,735 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+// SWD (Serial Wire Debug) control from the RoT to the SP
+//
+// The ARM Debug Interface Specification (ADI) describes the Serial Wire Debug
+// Port (SW-DP) protocol. This is frequently implemented as a bit-bang protocol
+// (setting the GPIO pins manually). Per 5.3.2 of ADIv5
+//
+// A successful write operation consists of three phases:
+// - an eight-bit write packet request, from the host to the target
+// - a three-bit OK acknowledge response, from the target to the host
+// - a 33-bit data write phase, from the host to the target.
+//
+// There is one bit of turnaround between the request and acknowledge and
+// the acknowledge and write phase.
+//
+// Per 5.3.3 of ADIv5
+//
+// A successful read operation consists of three phases:
+// - an eight-bit read packet request, from the host to the target
+// - a three-bit OK acknowledge response, from the target to the host
+// - a 33-bit data read phase, where data is transferred from the target to the host
+//
+// There is one bit of tunraound between the request and acknowledge.
+//
+// It turns out this specification can be implemented on top of a SPI block with
+// some specific implementation details:
+// - SPI has 4 pins (MOSI, MISO, CS, CSK) and importantly expects MOSI and MISO
+// to be separate. This is in contrast to SWD which assumes a single pin for
+// both input and output. For our SPI implementation, we tie MOSI and MISO
+// together and configure exactly one of MOSI or MISO for use with SPI at a time.
+// A side effect of this choice is that reading needs to be precise with the
+// specification. Adding extra idle cycles is fairly easy when writing but that
+// is not possible with reading.
+// - The LPC55 can transmit between 4 and 16 bits at a time. This makes
+// makes transmitting the various combinations of phases and turnaround bits
+// a bit of a pain. This is broken up into the following combinations:
+// -- 8-bit packet write
+// -- 4-bit ACK read (three bits of response + turnaround)
+// -- 34-bit write (32 bits data, one bit partiy, one bit turnaround) broken up
```

Comment:
There is no turnaround bit required at the end of a write, this should be 33 bits.